### DISCUSSION TITLE
Modifying the source code of OpenFAST to enable the use of Simulink for the active structural control

### DISCUSSION THREAD

**Discussion by relhamoud** [OpenFAST/openfast#2546] [Q&A]
Dear Jason JONKMAN,

I am writing to ask you about the coupling between OpenFAST and Simulink. In fact, for the timebeing, OpenFAST supports Simulink in ServoDyn module where Simulink and OpenFAST can exchange real time data related to blade pitch angle and/or generator torque.
However, OpenFAST does not support Simulink in the Structural Control module to account for active structural control.

So, i was wondering how to do this coupling between Simulink and OpenFAST for the active structural control of the 5MW barge-type FOWT? Of course, i need to change the source code of OpenFAST and recompile.

But, from where to start ? and what do you think ?

Thank you in advance.

Best Regards,

Riad

**Comment by jjonkman** [comment]
Dear @relhamoud,

Its sounds like you are trying to enable active StC control through Simulink, StC_CMODE = 4.  See the related questions on OpenFAST issues: [Add the control TMD to Simulink with S-function · Issue #761 · OpenFAST/openfast · GitHub](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOpenFAST%2Fopenfast%2Fissues%2F761&data=05%7C02%7Cjason.jonkman%40nrel.gov%7Ce1304ba57f354bc5a92d08dd093bcf4a%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C638676875233949509%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=XROMbAxAtUuIU2OzOR0ZePaoHGcjfueAI65DbycynN8%3D&reserved=0) and [How to implement active control of TMD in Simulink · Issue #2323 · OpenFAST/openfast · GitHub](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOpenFAST%2Fopenfast%2Fissues%2F2323&data=05%7C02%7Cjason.jonkman%40nrel.gov%7Ce1304ba57f354bc5a92d08dd093bcf4a%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C638676875233969624%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=%2BDO2iLpwJYOXtIvbo9xrMO0kW1ZhInuhz%2BdRElBYGk0%3D&reserved=0).

Best regards,
